File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/GUI/MobzyGUI.java
package com.offz.spigot.mobzy.GUI;

import com.derongan.minecraft.guiy.gui.*;
import com.offz.spigot.mobzy.Builders.MobBuilder;
import com.offz.spigot.mobzy.GUI.Layouts.MobConfigLayout;
import com.offz.spigot.mobzy.Mobzy;
import com.offz.spigot.mobzy.MobzyAPI;
import com.offz.spigot.mobzy.Spawning.SpawnRegistry;
import de.erethon.headlib.HeadLib;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

public class MobzyGUI extends GuiHolder {
    private Player player;
    //    private Layout spawns;
    private Mobzy plugin;
    private List<Layout> history = new ArrayList<>();
    private ClickableElement back;
    private List<ClickableElement> mobConfigs = new ArrayList<>();
    private List<ClickableElement> spawnList = new ArrayList<>();
    private FileConfiguration config;

    public MobzyGUI(Player player, Mobzy plugin) {
        super(6, "Mobzy", plugin);
        this.plugin = plugin;
        this.player = player;

        //create back button
        Element cell = Cell.forItemStack(HeadLib.RED_X.toItemStack("Back"));
        back = new ClickableElement(cell);
        back.setClickAction(clickEvent -> backInHistory());

//        spawns = buildSpawns();
//        regionLayout = buildRegions();
        setElement(buildMobConfigLayout());
    }

    public void addBackButton(Layout layout) {
        history.add(layout);
        layout.addElement(8, 5, back);
    }

    public void backInHistory() {
        if (history.size() <= 1) {
            player.closeInventory();
            return;
        }
        setElement(history.get(history.size() - 2));
        history.remove(history.size() - 1);
    }

    private Layout buildMobConfigLayout() {
        Layout layout = new Layout();

        FillableElement grid = new FillableElement(4, 8);
        layout.addElement(0, 0, grid);

        player.sendMessage(plugin.getMobzyConfig().getMobCfgs().values() + "");
        Collection<FileConfiguration> values = plugin.getMobzyConfig().getSpawnCfgs().values();
        if (values.size() == 1)
            return buildRegions(values.iterator().next());
        else
            values.forEach(config -> {
                String name = "Unnamed configuration";
                Material icon = Material.BEDROCK;

                if (config.contains("config.name"))
                    name = config.getString("config.name");
                if (config.contains("config.icon"))
                    icon = Material.getMaterial(config.getString("config.icon"));

                Element cell = Cell.forMaterial(icon, name);
                ClickableElement mobConfig = new ClickableElement(cell);
                mobConfig.setClickAction(clickEvent -> setElement(buildRegions(config)));

                grid.addElement(mobConfig);
            });

        addBackButton(layout);
        return layout;
    }

    private Layout buildRegions(FileConfiguration config) {
        this.config = config;
        Layout layout = new Layout();

        FillableElement grid = new FillableElement(4, 8);
        layout.addElement(0, 0, grid);

        config.getMapList("regions").forEach(region -> {
            String regionName = (String) region.get("name");
            player.sendMessage(regionName);

            Element cell = Cell.forMaterial(Material.DIAMOND, regionName);
            ClickableElement mobConfig = new ClickableElement(cell);
            mobConfig.setClickAction(clickEvent -> setElement(buildSpawns((ArrayList<Map<String, Object>>) region.get("spawns"), (String) region.get("name"))));
            grid.addElement(mobConfig);
        });

        addBackButton(layout);
        return layout;
    }

    private Layout buildSpawns(ArrayList<Map<String, Object>> spawns, String regionName) {
        Layout layout = new Layout();

        FillableElement grid = new FillableElement(4, 8);
        layout.addElement(0, 0, grid);

        spawns.forEach(spawn -> {
            MobBuilder spawnBuilder;
            if (spawn.containsKey("reuse"))
                spawnBuilder = MobzyAPI.getBuilder(SpawnRegistry.getReusedBuilder((String) spawn.get("reuse")).getEntityType().c().getSimpleName());
            else {
                String mobName = (String) spawn.get("mob");
                player.sendMessage(mobName);
                spawnBuilder = MobzyAPI.getBuilder(mobName);
            }
            Element cell = Cell.forItemStack(spawnBuilder.getModelItemStack());
            ClickableElement mobConfig = new ClickableElement(cell);
            //open up the config layout with its menu options
            mobConfig.setClickAction(clickEvent -> setElement(new MobConfigLayout(this, spawn, regionName)));
            grid.addElement(mobConfig);
        });

        addBackButton(layout);
        return layout;
    }

    public void saveConfigValues(Map<String, Object> spawn, List<MobzyPropertyElement> mobProperties) {
        spawn.clear();
        mobProperties.forEach(property -> spawn.put(property.getKey(), property.getValue()));
        plugin.getMobzyConfig().saveSpawnCfg(config);
        player.sendMessage(ChatColor.GREEN + "Successfully saved mob's configuartion");
    }
}

File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/gui/MobzyGUI.kt
package com.offz.spigot.mobzy.gui

import com.derongan.minecraft.guiy.gui.Cell
import com.derongan.minecraft.guiy.gui.ClickableElement
import com.derongan.minecraft.guiy.gui.FillableElement
import com.derongan.minecraft.guiy.gui.Layout
import com.derongan.minecraft.guiy.gui.layouts.HistoryGuiHolder
import com.offz.spigot.mobzy.gui.layouts.MobConfigLayout
import com.offz.spigot.mobzy.mobTemplate
import com.offz.spigot.mobzy.mobzy
import com.offz.spigot.mobzy.spawning.SpawnRegistry.reuseMobSpawn
import com.offz.spigot.mobzy.toTemplate
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.entity.Player
import java.util.*
import java.util.function.Consumer

//TODO cleanup code after conversion to kotlin
class MobzyGUI(val player: Player) : HistoryGuiHolder(6, "Mobzy", mobzy) {
    private val mobConfigs: List<ClickableElement> = ArrayList()
    private val spawnList: List<ClickableElement> = ArrayList()
    private var config: FileConfiguration? = null

    private fun buildMobConfigLayout(): Layout {
        val layout = Layout()
        val grid = FillableElement(4, 8)
        layout.addElement(0, 0, grid)
        val configs: Collection<FileConfiguration> = mobzy.mobzyConfig.spawnCfgs.values
        if (configs.size == 1) return buildRegions(configs.first())
        else configs.forEach { config ->
            val name = if (config.contains("config.name")) config.getString("config.name")!! else "Unnamed configuration"
            val icon = if (config.contains("config.icon")) Material.getMaterial(config.getString("config.icon")!!)!! else Material.BEDROCK
            val cell = Cell.forMaterial(icon, name)
            val mobConfig = ClickableElement(cell) { setElement(buildRegions(config)) }
            grid.addElement(mobConfig)
        }
        addBackButton(layout)
        return layout
    }

    private fun buildRegions(config: FileConfiguration): Layout {
        this.config = config
        val layout = Layout()
        val grid = FillableElement(4, 8)
        layout.addElement(0, 0, grid)
        config.getMapList("regions").forEach(Consumer { region: Map<*, *> ->
            val regionName = region["name"] as String?
            var material: Material? = Material.BEDROCK
            if (region.containsKey("icon")) {
                val tempMaterial = Material.getMaterial((region["icon"] as String?)!!)
                if (tempMaterial != null) material = tempMaterial
            }
            val cell = Cell.forMaterial(material, regionName)
            val mobConfig = ClickableElement(cell) {
                @Suppress("UNCHECKED_CAST")
                setElement(buildSpawns(region["spawns"] as List<MutableMap<String, Any?>>, region["name"] as String))
            }
            grid.addElement(mobConfig)
        })
        addBackButton(layout)
        return layout
    }

    private fun buildSpawns(spawns: List<MutableMap<String, Any?>>, regionName: String): Layout {
        val layout = Layout()
        val grid = FillableElement(4, 8)
        layout.addElement(0, 0, grid)
        spawns.forEach { spawn ->
            val spawnBuilder =
                    if (spawn.containsKey("reuse"))
                        reuseMobSpawn(spawn["reuse"] as String).entityType.mobTemplate //TODO not sure if this is right because I changed it from before
                    else
                        (spawn["mob"] as String).toTemplate()
            val cell = Cell.forItemStack(spawnBuilder.modelItemStack)
            //open up the config layout with its menu options
            val mobConfig = ClickableElement(cell) { setElement(MobConfigLayout(this, spawn, regionName)) }
            grid.addElement(mobConfig)
        }
        addBackButton(layout)
        return layout
    }

    fun saveConfigValues(spawn: MutableMap<String, Any?>, mobProperties: List<MobzyPropertyElement>) {
        spawn.clear()
        mobProperties.forEach { spawn[it.key] = it.value }
        mobzy.mobzyConfig.saveSpawnCfg(config ?: error("Could not save config values, config was null"))
        player.sendMessage("${ChatColor.GREEN}Successfully saved mob's configuartion")
    }

    init {
        setElement(buildMobConfigLayout())
    }
}